Add Google App Engine Testing Instance support (with RequestFactory interface{})

mattes opened the pull request:
This PR adds support for Google App Engine Testing.

#### Background
Each `*http.Request` needs to be registered with the GAE test instance. Otherwise calls to `appengine.NewContext(*http.Request)` will panic with:

```
appengine: NewContext passed an unknown http.Request
```

#### References
* https://github.com/golang/appengine/blob/master/aetest/instance_vm.go#L60
* https://cloud.google.com/appengine/docs/go/tools/localunittesting/

#### Changes

* Add `GaeTestInstance` to `struct Config`
* If `Config.GaeTestInstance` is specified (not nil), use `GaeTestInstance.NewRequest` to create `*http.Request` that is associated with the GAE test instance.
* Otherwise create plain `*http.Request`.
* In order to keep the correct reference, `struct Request` needs pointer to request.

mattes commented:
This works for me so far. `_examples/gae.go` is more like an example usage, not actual tests. Tests would require the Google App Engine Go SDK. 

Leaving this PR here for discussion and feedback. Happy to change things if necessary. Once finalized, I will change the import paths, too.

mattes commented on _examples/gae.go:
```diff
@@ -0,0 +1,43 @@
+	return nil
+}
+
+func NewHttpExpect(t *testing.T) *httpexpect.Expect {
```
I wonder if this could be a "first class" func, like 

```
func NewGae(t *testing.T, r http.Handler, i aetest.Instance) *httpexpect.Expect
```
Thoughts?

gavv commented:
Hi, thanks for the PR!

I'd prefer to hide request creation behind an interface, say:

```
type RequestFactory interface {
    NewRequest(method, urlStr string, body io.Reader) (*http.Request, error)
}

...

type Config struct {
    ...
    RequestFactory RequestFactory
}
```

and use it in request.go instead of GaeTestInstance.

This interface is compatible with the aetest.Instance, thus the user can just pass the instance as a factory.

If the user didn't specify a factory, the default RequestFactory implementation should be used (say, DefaultRequestFactory) that just calls http.NewRequest.

With this approach, we can also eliminate aetest dependency.

gavv commented:
Meanwhile, we can still keep aetest dependency in examples.

gavv commented:
> I wonder if this could be a "first class" func, like 
> [...]
> Thoughts?

We could a more generic wrapper:

```
func WithHandler(t LoggerReporter, r http.Handler, f RequestFactory) *httpexpect.Expect
```

mattes commented:
Very good thoughts. I will update that.

mattes commented:
* Changed everything to use a `RequestFactory` interface
* Added tests
* Updated `_examples/gae_test.go` to include fully working example 
* Couldn't decide on the `func WithHandler(t LoggerReporter, r http.Handler, f RequestFactory) *httpexpect.Expect` idea, so left it out for now. Might open a new PR for it in the future. 

I think my work is done here :-)

gavv commented:
Great.

This PR had two issues:

* the Request struct was not prepared to handle nil http.Request, which is now possible since RequestFactory can fail

* it's not correct to use "path" as an URL in NewRequest, since it may contain `{named}` arguments for go-interpol that will be called later from WithPath

PR is merged and this two issues are fixed as well.

I've also extracted DefaultRequestFactory struct from NewRequest and edited README a bit.

mattes commented:
Awesome! Thanks for fixing the issues and merging! One of the cleanest code bases I've seen in a while. Keep it up!
